src/main/scala/tech/cryptonomic/conseil/generic/chain/MultiFetchDecoding.scala

```diff
@@ -0,0 +1,136 @@
+package tech.cryptonomic.conseil.generic.chain
+
+import cats._
+import cats.data.Kleisli
+import cats.syntax.semigroupal._
+
+/** Contract for a generic combination of operations that get encoded data (usually json) from
+  * some resource (e.g. the tezos node), then converts that to a value, batching multiple requests together.
+  * The returned values are tupled with the corresponding input
+  * The interface assumes some kind of higher-order typed effect `F[_]` is part
+  * of the operation execution, and represents everything parametrically, to leave
+  * room for future evolutions of the code protocol.
+  * Thus the required implementation is in the form of a `cats.data.Kleisli` type,
+  * parametric on effect `F`, input `In`, output `Out`. As needed, the in/out types are
+  * wrapped in a `T[_]` type constructor that represents multiple values (e.g. a collection type).
+  *
+  * @tparam F the higher-order effect type (e.g. `Try`, `Future`, `IO`, `Either`)
+  * @tparam T a traversable type (used to wrap mutliple requests and corresponding return values)
+  */
+trait MultiFetchDecoding[F[_], T[_]] {
+  /** the input type, e.g. ids of data */
+  type In
+  /** the output type, e.g. the decoded block data */
+  type Out
+  /** the encoded representation type used e.g. some Json representation */
+  type Encoded
+
+  /** an effectful function from a collection of inputs `T[In]`
+    * to the collection of encoded values, tupled with the corresponding input `T[(In, Encoded)]`
+    */
+  def fetchBatch: Kleisli[F, T[In], T[(In, Encoded)]]
+
+  /** an effectful function that decodes the json value to an output `Out` */
+  def decodeData: Kleisli[F, Encoded, Out]
+
+  /** using the combination of the provided functions we fetch all data with this general outline:
+    * 1. get json values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and convert every json to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    */
+  def fetch(implicit app: Monad[F], tr: Traverse[T]): Kleisli[F, T[In], T[(In, Out)]] =
+    fetchBatch.andThen(json => decodeData.second[In].traverse(json))
+
+}
+
+object MultiFetchDecoding {
+  import cats.syntax.foldable._
+  import cats.syntax.functorFilter._
+
+  /* An alias used to define function constraints on the internal dependent types (i.e. `I`, `O`, `E`)
+   * which would be otherwise opaque in any function signature
+   */
+  private type Aux[F[_], T[_], I, O, E] = MultiFetchDecoding[F, T] {
+      type In = I
+      type Out = O
+      type Encoded = E
+    }
+
+  /** Adds an extra decoding operation for every intermediate encoded result, returning both the
+    * original output `O` and the new decoded output `O2` as a tuple
+    */
+  def addDecoding[F[_]: Apply, T[_], I, O, O2, E](
```

Comment:
For this function and the ones below, could you add some concrete examples of how they would be useful for fetching blockchain data?

Reply:
This is a wonderful suggestion, I'll add examples to all cases